lib/auth/grpcserver.go

```diff
@@ -1612,26 +1627,117 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
 			return trace.BadParameter("missing NodeName field in a ssh-only UserCertsRequest")
 		}
+		// Find the target node and check whether MFA is required.
+		nodes, err := actx.authServer.GetNodes(defaults.Namespace, services.SkipValidation())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		var matches []types.Server
+		for _, n := range nodes {
+			// Get the server address without port number.
+			addr, _, err := net.SplitHostPort(n.GetAddr())
+			if err != nil {
+				addr = n.GetAddr()
+			}
+			// Match NodeName to UUID, hostname or self-reported server address.
+			if n.GetName() == req.NodeName || n.GetHostname() == req.NodeName || addr == req.NodeName {
+				matches = append(matches, n)
+			}
+		}
+		if len(matches) == 0 {
+			return notFoundErr
+		}
+		if len(matches) > 1 {
+			return trace.BadParameter("multiple nodes match %q, please use the node UUID to login instead", req.NodeName)
+		}
+		noMFAAccessErr = actx.context.Checker.CheckAccessToServer(req.Username, matches[0], false)
+
 	case proto.UserCertsRequest_Kubernetes:
 		if req.KubernetesCluster == "" {
 			return trace.BadParameter("missing KubernetesCluster field in a kubernetes-only UserCertsRequest")
 		}
+		// Find the target cluster and check whether MFA is required.
+		svcs, err := actx.authServer.GetKubeServices(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		var cluster *types.KubernetesCluster
+	outer:
+		for _, svc := range svcs {
+			for _, c := range svc.GetKubernetesClusters() {
+				if c.Name == req.KubernetesCluster {
+					cluster = c
+					break outer
+				}
+			}
+		}
+		if cluster == nil {
+			return notFoundErr
+		}
+		noMFAAccessErr = actx.context.Checker.CheckAccessToKubernetes(defaults.Namespace, cluster, false)
+
 	case proto.UserCertsRequest_Database:
 		if req.RouteToDatabase.ServiceName == "" {
 			return trace.BadParameter("missing ServiceName field in a database-only UserCertsRequest")
 		}
+		dbs, err := actx.authServer.GetDatabaseServers(ctx, defaults.Namespace, services.SkipValidation())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		var db types.DatabaseServer
+		for _, d := range dbs {
+			if d.GetName() == req.RouteToDatabase.ServiceName {
+				db = d
+				break
+			}
+		}
+		if db == nil {
+			return notFoundErr
+		}
+		noMFAAccessErr = actx.context.Checker.CheckAccessToDatabase(db, false)
+
 	case proto.UserCertsRequest_All:
 		return trace.BadParameter("must specify a concrete Usage in UserCertsRequest, one of SSH, Kubernetes or Database")
 	default:
 		return trace.BadParameter("unknown certificate Usage %q", req.Usage)
 	}
-	maxExpiry := clock.Now().Add(teleport.UserSingleUseCertTTL)
+	// No error means that MFA is not required for this resource by
+	// AccessChecker.
+	if noMFAAccessErr == nil {
+		return errUserSingleUseUserCertNotNeeded
+	}
+	// Errors other than ErrSessionMFARequired mean something else is wrong,
+	// most likely access denied.
+	if noMFAAccessErr != services.ErrSessionMFARequired {
```

Comment:
good point.
I used `trace.Unwrap` originally and had trouble (because I think `trace.AccessDenied` unwraps into the underlying string error). But `errors.Is` works well, updated.